feat: GitHub CLI authentication support

lox opened the pull request:
Adds support for using GitHub CLI (`gh`) as an authentication provider for Hermit,  eliminating the need to manually set `HERMIT_GITHUB_TOKEN`.

Exposing `HERMIT_GITHUB_TOKEN` securely across many developer environments is tricky to do safely without exposing it to other tools or distributing shell wrappers for hermit. Using `gh auth token` allows for short-lived tokens that are only exposed to hermit as needed. 

This feature is enabled via setting the following in the user hermit config (~/.hermit.hcl):

```
gh-cli-auth = true;
```

lox commented:
This was tricky to test! 

What I've done is add a build tag of `localgithub` that is used by the integration tests to compile hermit. If compiled this way, it respects `HERMIT_GITHUB_BASE_URL` as an alternative to `github.com` which allows tests to create a mock github endpoint without the horrors of HTTPS proxying.

Lemme know if this is overkill @alecthomas.

nickajacks1 commented on github/auth/provider.go:
```diff
@@ -0,0 +1,88 @@
+		return "", errors.Wrap(err, "failed to get token from gh CLI")
+	}
+
+	p.ui.Tracef("gh auth token output: %s", string(output))
```
Hmmm, this prints a secret. It's probably better to either remove this or warn in the cli usage that trace may leak sensitive information.

lox replied:
Good catch, was there for dev.

nickajacks1 reviewed (commented):
Seems good to me. Just had one more little comment.

lox commented on app/main.go:
```diff
@@ -217,13 +227,36 @@ func Main(config Config) {
+		}
+
+		provider, err := auth.NewProvider(providerType, p)
+		if err != nil {
```
Fixed, both are now Fatal.

nickajacks1 reviewed (approved):
Not a maintainer but looks LGTM to me

lox commented:
Appreciate the review @nickajacks1

alecthomas reviewed (approved):
I like it!

lox committed:
feat: add GitHub CLI authentication support

Add support for using GitHub CLI (`gh`) as an authentication provider for Hermit via a user config (`gh-cli-auth`), enabling seamless integration with `gh auth token`.
This simplifies authentication for private packages and repositories by eliminating the need to manually set HERMIT_GITHUB_TOKEN.

lox committed:
Remove token debug logging

lox committed:
Use constants for provider types

lox committed:
Defer auth token fetching to when called

lox committed:
auth.NewProvider or GetToken() failure should be fatal

lox committed:
fix: restore cli setup before github auth

lox committed:
Test failed gh auth token call
